### ISSUE TITLE
RFv2 speed

### ISSUE THREAD

**Issue by itwysgsl** [bschn2/rainforest#15]
Hello again @bschn2, I have concerns about v2 speed. Don't you think that rainforest became way to slow? My MacBook's Intel i7 produce ~30 H/s and it's not enough to mine at least one block on lowest diff on test network 😅

**Comment by wtarreau** [comment]
Strange about these 30 H/s, it's 40 times slower than my i7 (mine does 1200). Did you build with -march=native -O3 ?
How did you test, using rf_test or did you simply iterate over rh_hash() with a NULL rambox ? This last one will solely depend on memory speed since it needs to initialize it. That's what I figured when implementing the bench mode in rf_test. You have to pre-allocate a rambox and pass it to rf_hash(), which restores it at the end. At first I didn't like the principle but thinking about the reasons (making it very hard for FPGAs and ASICs to try to emulate the RAM) I finally found it very smart, as the cost is only incurred to those trying to massively scale it ;-)

**Comment by wtarreau** [comment]
My old ATOM D510 does 55 H/s :-)

**Comment by itwysgsl** [comment]
@wtarreau oh, I guess there is little missunderstanding. It gives me ~30 H/s per thread using cpu miner source code example from rfv2 branch. I just tried recompile test example with `-march=native -O3` option and now it give me ~ 130 H/s per thread :D

**Comment by itwysgsl** [comment]
Interesting, what purpose of this `-march=native -O3` option?

**Comment by wtarreau** [comment]
On Thu, Apr 04, 2019 at 11:48:04PM -0700, iamstenman wrote:
> @wtarreau oh, I guess there is little missunderstanding. It gives me ~30 H/s
> per thread using cpu miner source code example from rfv2 branch. I just tried
> recompile test example with `-march=native -O3` option and now it give me ~
> 130 H/s per thread :D 

Much better!

> Interesting, what purpose of this `-march=native -O3` option?

-O3 is the optimisation level. It's the highest on most compilers.
Recently gcc introduced -Ofast which is more or less the same.
-march=native indicates that certain features present on the CPU
you're building on will be enabled. This is recommended when you
don't want to manually enable/disable each and every feature and
you know you're building to run on your local machine.

You can check the output of these :

   $ gcc -dM -xc -E -< /dev/null | less

   $ gcc -march=native -dM -xc -E -< /dev/null | less

-dM dumps all known macros
-xc says the source is a C file
-E says just emit the preprocessed output
So you end up with a lot of defines.

For example you'll find some AES, SSE, AVX and whatever is available
on *this* CPU that is not available on another one. This is why my
server's Atom was dying in "illegal instruction" previously when I
tried to run it from the executable made for the i7 :-)

Willy

**Comment by itwysgsl** [comment]
Hello again @wtarreau. After some tests (~20 minutes mining on testnet on lowest difficulty without any blocks) I still think that RFv2 in current state way to slow :(

**Comment by wtarreau** [comment]
On Fri, Apr 05, 2019 at 06:04:18PM +0000, iamstenman wrote:
> Hello again @wtarreau. After some tests (~20 minutes mining on testnet on
> lowest difficulty without any blocks) I still think that RFv2 in current
> state way to slow :(

It's very possible, I don't know much what all this implies in fact :-)
Probably that lowering the number of rounds would help a lot, though
maybe some other parts of the algo are very expensive as well. What
would be an correct hashrate in your opinion ? I'm interested by pure
curiosity in checking on various devices (arm, x86) the impacts of
various knobs.

Willy

**Comment by wtarreau** [comment]
For example, I just checked and just changing this multiples the perf by
3 :

-       if (__builtin_clrsbl(old) > 5) {
+       if (__builtin_clrsbl(old) > 4) {

(...)

-       loops = sin_scaled(msgh) * 3;
+       loops = sin_scaled(msgh) + 1;

Maybe that's enough in your case ?

Willy

**Comment by wtarreau** [comment]
On Fri, Apr 05, 2019 at 09:34:08PM +0200, Willy Tarreau wrote:
> For example, I just checked and just changing this multiples the perf by
> 3 :
> 
> -       if (__builtin_clrsbl(old) > 5) {
> +       if (__builtin_clrsbl(old) > 4) {
> 
> (...)
> 
> -       loops = sin_scaled(msgh) * 3;
> +       loops = sin_scaled(msgh) + 1;
> 
> Maybe that's enough in your case ?

I can even go further by removing two sets of divbox+scramble in
the round. My i7 reaches 4500 H/s there.

Willy

**Comment by wtarreau** [comment]
And I can reach 6k hashes/s on the i7 by flattening the curve giving the
number of loops to reduce the extremes :

 static uint8_t sin_scaled(unsigned int x)
 {
-       return pow(sin(x / 16.0), 5) * 127.0 + 128.0;
+       return sqrt(sqrt(pow(sin(x / 16.0), 5) + 1.0)) * 100.0;
 }
 
(rpi does 3900 here and npi 10300).

Willy

**Comment by LinuXperia** [comment]
I was never a fan of this difficulty target calculation idea.

The whole Bitcoin difficulty target calculation is a second class solution and does not really works.

Imagine the hashrate goes very high and the difficulty increases to maximum.
Then only 1 Hash or a very small number of hashes can be below such a extreme set Target
and producing such a Hash may even not be possible as no data in the block header can produce such a hash.

It is highly questionable even if it will be possible to create the needed hashes for MicroBitcoin as 21 trilions of coins need to be mined instead the 21 million coins bitcoin has.

We may very well get into problems with not having enogh of hashes to be mined in MicroBitcoin as trillions of Coins are planned to be mined.

Becouse of this lot of wasted hash calculation are produced and energy is wasted.

Reducing the Loops as i see benefits big Power Hungry CPUs over small power efficient CPUs and is against the Idea of the RainForest Hash Algorithm.

I suggest to go the path that EquiHash uses and is implemented as a example in Bitcoin Gold or ZCash and instead of finding a hash that is below
a target finding instead half hashes colusions inside the rambox.

I really recommend to drop this whole target calculation and instead use
the EquiHash way to find colusions inside the Rambox like Equihash does it.

If my memory serves me right EquiHash creates 1 RamBox
and does populate the RamBox for every Nonce using 1024 Loops with Salsa20 calculated Hashes.

After this it looks for Hash Colusions inside this RamBox and if it finds it
append it after the Nonce in the Blockheader and submit it to the Network.

Becouse of this the Block Header of EquiHash has 80 Bytes plus the 32 Byte Hash colusion hash.

Maybe something simillar can be done also with the RainForest Hash Algorithm ?

**Comment by bschn2** [comment]
Gentlemen,

please be careful if you start to change the number of loops, iterations per round, or the loop curve!

@wtarreau at least make sure you round the pow up by adding 1.5 and not 1.0. This aside, your suggestion looks reasonable to me as it gives more a body-like shape which stays longer in the higher range hence reduces the peaks. In any case you must check the average number of history buffer entries and double it in the define (well maybe less than double now with the curve change but it must at least be 20-25% larger to avoid repopulating the rambox from sratch at the end). If the number of entries is lower than 512 then you need to decrease the clrsbl threshold so that it writes more often.

The divbox+scramble calls you suggested to remove were indeed here only to balance the power between low-end CPUs and high-end ones. There are exactly 3 which are doubled and could safely be reduced to 1 each (the original smhasher tests were run with both configurations).

And yes, please keep an eye on raspberry-pi type of devices as it seems capital to me that such machines are about as fast as regular PCs if we really want to incentive to energy savings. Your numbers are fine to me as I initially targetted 1k to 10k H/s/core so we're pretty much in this area here.

**Comment by jdelorme3** [comment]
And how many times more faster it is on large devices and small devices for equihash ? Are you sure we  don't favour the large ones only there ?

**Comment by jdelorme3** [comment]
Also the hash verifycation time count alot. I think that rainforest is great for this, it cost but not way two much.

**Comment by bschn2** [comment]
Dear @LinuXperia,

rfv2 's rambox is not far from what you describe since the rambox is modified by every lookup based on the hashed message (thus includes the nonce). However keep in mind that Salsa20 was designed to be extremely fast on x86 processors (typically less than 4 cycles per byte), and that this can hardly be considered fair for emerging countries where such hardware simply is not available, all people have is a previous generation smartphone to do everything. There the power often comes from local solar panels so maintaining a decent capacity to such devices is very important for the overall fossil energy consumption.

**Comment by wtarreau** [comment]
@bschn2 good catch for 1.5, I'll try this to stay safe. Thanks for confirming the divbox that could be reduced. Regarding the write ratio if you notice I already adjusted it, but granted, i didn't check the values. I'll do and will prepare a patch will all this soon.

**Comment by LinuXperia** [comment]
@bschn2 
I am sorry it looks like i did not express my self very good how to improve the actual situation with Rainforest so that we instead look to find a rare Hash with leading Zeros we look to find a collusion of each calculated Nonce Hash in the Rambox.

here is my easy to implement improvement suggestion:

The Problem is the part that require us to check for a specific rare hash with leading zeros in front of the Hash.
Such a rare hash occurence requires a lot of brute force hash calculation which is against what rainforest algorithm stands for.

My working approach that solves this problem without changing a lot of the code is that instead we look to find a rare end hash with leading zeros  in front and be under a target hash we use the amount of the leading zeros in the nBits field value as amount of leading bytes of the calculated Nonce endhash to be matched in the rambox.

Lets say the lowest dificlulty require us to have a hash with two leading zeros. 
Instead that we now do hash bruteforce till a rare hash is found with two leading zeros what we do now with rainforest is we match two bytes of the calculated end hash and look if such a byte variation collusion exist in the rambox using memcmp().

The Value of how many leading Zeros are required is stored in the 80 Bytes pdata Blockheader that each thread has. So implementing this is very easy and should work like a charme.

If somebody finds such a Byte End Hash Collusion very fast then the difficulty adjust automaticly the nbits field and require us to find a hash with let say 8 leading zeros now which is more harder than before.
 
We again instead looking to find a endhash with 8 leading zeros with rainforest just match 8 bytes of each nonce end hash and look in the the rambox if such a Byte combination exist.

If yes and this was found again faster than the 1 Minute requirement that Microbitcoin has for mining a block the difficulty Algorithm will adjust the nbits field value to find a hash with lets say 16 leading Zeros which for us and the Rainforest algoirthm means we need to map 16 Bytes of each calculated Nonce endhash in the rambox. 

If finding such a 16 Byte Hash Combination needs now 2 Minutes instead the required 1 Minute then the dificulty algotithm will drop the difficulty in the nbits field to look for 14 leading zeros aka 14 Bytes in the rambox and make it easier than before. 

By this it automaticly adjust everything so we stay inside the 1 Minute Time frame for mining a Block without that we need to bruteforce Hash calculations to find a rare hash with leadin zeros.

**Comment by itwysgsl** [comment]
Hello again @bschn2 @wtarreau.
I just tested this https://github.com/bschn2/rainforest/commit/3b35a37990a546856953566cd967b35b90d27733 commit and it's actually ~3-4 times faster on my i7. But it's still not fast enough. I tried to mine around 10 minutes on same testnet with low diff and etc, but it still the same. After that I started experimenting by dividing amout of loops at this line for 3, 6, 12 and so on (don't ask why, I just wanted to test speed 😅): https://github.com/bschn2/rainforest/blob/3b35a37990a546856953566cd967b35b90d27733/rfv2_core.c#L726 
And here is what I got:
- Divided by 3: 1111.15 H/s/thread without any blocks in a long time
- Divided by 6: 2533.53 H/s/thread 
- Divided by 12: 5446.98 H/s/thread 
- Divided by 24: 11606.02 H/s/thread at this point I start getting blocks with more or less decent speed

Maybe this test would be helpful in some way :)

**Comment by bschn2** [comment]
@itwysgsl do you mean you don't find shares or you don't find blocks ? Not finding shares is indeed problematic but should simply be a matter of difficulty. Indeed, at 1.11kH/s you scan a full 16-bit range every 60 seconds so the pool's difficulty is low enough, you must find these shares. What is your difficulty in this case ?

If what you don't find is a block, this sounds normal as the purpose is that the chances to find a block are equally shared among miners so if you have 1000 miners one will mine the block while the 999 other ones will not. So on average if you emit a new block every minute, for 1000 miners each of them would on average find a block every 1000 minutes. But again even then it's a matter of adjusting the difficulty as if the target is 0x0000FFFF...FFFF then at 1kH/s you *will* find it in 60 seconds.

Last point, I'm a bit surprised by your i7's performance here, did you enable correct build options ? This is roughly 5 times slower than mine without dividing, hence 15 times slower. Did you enable -O3 and -march=native ?

**Comment by bschn2** [comment]
Oh and by the way, many thanks for sharing your observations!

**Comment by bschn2** [comment]
@LinuXperia I'm still unsure I really understand the principle you're describing. I think it is very similar to hashing except that you look up some bits in the rambox. What I don't understand is how you populate it and how you validate a share or a block afterwards. Also in any case the computation time spent is required as a proof of work. Whether you find the bits in the rambox or anywhere else inside the hash algorithm, it's the same, you have to iterate over nonces so that most participants find shares to be paid and that one of them finds the block. So it's unclear to me what your method brings at this point.

**Comment by wtarreau** [comment]
@itwysgsl I am also surprised by your numbers, how do you test ? Is it with the patched cpuminer maybe ? Have you tried "rfv2_test -b -t $(nproc)" ? I must confess I have not checked how or when it initializes the rambox. I hope it does it only once once the first call, but I don't know. This could explain your low performance if it does a full rambox for each hash.

**Comment by LinuXperia** [comment]
> @LinuXperia Whether you find the bits in the rambox or anywhere else inside the hash algorithm, it's the same, you have to iterate over nonces so that most participants find shares to be paid and that one of them finds the block. So it's unclear to me what your method brings at this point.

@bschn2 
The way i understand it is that the 110 H/s/Thread are way to low to find a Block in about 60 Seconds as a Single Miner at the lowest difficulty.

Becouse of this Problem improvements are needed so a Single Miner using a Single Board Computer running Microbitcoin Rainforest Miner is able to mine a Block as Solo Miner in about 60 Seconds.
The test of just a single miner runing just one node and one Miner failed as he was not able to Mine any blcoks in the given time. 

So the minimal requirement to test on one Block Chain Node using one Miner to mine one Block in the given period failed.

His hashing Numbers looks okey as he has the same Hash Speed on his i7 as i have on my i7

Becouse of this i suggested to abdon the bitcoin aproach looking to find a rare hash with leading zeros and instead use the equihash apraoch in finding bit collusions.

Finding bit collusions make it easier to mine blocks as we dont need to bruteforce a huge amount of hashes and loose time till we found such a rare hash.

**Comment by wtarreau** [comment]
@LinuXperia how do you measure this performance, and how is this lowest difficulty calculated or configured (sorry I'm not much aware of all this, I'm only using cpuminer to validate the thermal robustness of my build farm). With rfv2_test I'm seeing numbers 8 times larger than yours:
$ gcc -march=native -O3 -o rfv2_test rfv2_test.c -pthread -lm
$ ./rfv2_test -b -t 1
847 hashes, 1.021 sec, 1 thread, 829.374 H/s, 829.374 H/s/thread
860 hashes, 1.000 sec, 1 thread, 859.975 H/s, 859.975 H/s/thread
849 hashes, 1.000 sec, 1 thread, 848.988 H/s, 848.988 H/s/thread
858 hashes, 1.000 sec, 1 thread, 857.988 H/s, 857.988 H/s/thread
^C
And on ARM:
$ ./rfv2_test -b -t 1
1334 hashes, 1.071 sec, 1 thread, 1245.832 H/s, 1245.832 H/s/thread
1336 hashes, 1.000 sec, 1 thread, 1335.768 H/s, 1335.768 H/s/thread
1333 hashes, 1.000 sec, 1 thread, 1332.841 H/s, 1332.841 H/s/thread
^C

**Comment by bschn2** [comment]
@LinuXperia well, I really don't understand the method you're trying to explain, I'm sorry. I don't understand why you say "rare hash with leading zeroes", the number of zeroes is log2(1/frequency) so if a matching hash is rare it's because it has been made so by the difficulty.
I will have a look at equihash to try to understand how it differs regarding this, but I still fail to see how that would change anything given that we *want* a miner to spend time to prove his work.

**Comment by bschn2** [comment]
Well after having read a bit about equihash I think I get it a bit better, but in my humble opinion it focuses solely on the memory-bound aspect and as a result it has already been ported to an ASIC (bitmain's Z9 which is 10 times faster for the price than a GPU) : https://www.heise.de/newsticker/meldung/Ende-der-Grafikkarten-Aera-8000-ASIC-Miner-fuer-Zcash-Bitcoin-Gold-Co-4091821.html
This even resulted in a 51% attack on Bitcoin Gold and a loss of $18M. This is exactly the type of things I want to avoid.

Also looking at the numbers, it's said that an Nvidia 1080Ti does only 650 sol/s (=hashes/s) so it's even way lower than what we're doing on rfv2. The main challenge we have to address is to make sure that MBC's short-lived blocks can be mined in the block's life, and the solution above apparently makes this situation worse from what I'm reading.

**Comment by endlessloop2** [comment]
Hello y'all, I've been following RF since last year and implemented V1 on my unreleased coin. 
I'd like to know what is the purpose of making the algorithm "faster", considering coins can change starting difficulty. 
What kind of tests are you doing that you are finding it "slow to find blocks"?
I believe the algorithm is fine at this point except for any bugs that may be unsolved, but it shouldn't be changed to make it "faster".

**Comment by wtarreau** [comment]
Interesting. It's important to keep in mind that memory speed varies with the device's price. The DRAM access times I've measured so far: http://git.1wt.eu/web?p=ramspeed.git;a=blob;f=data/results.txt
So a cheap board has 2-3 times the access time of a PC, and it's said (though I cannot verify) that GPUs are even faster with GDDR5. Also the PC's memory controller allows to initiate multiple accesses at once while the cheap devices don't, resulting in almost a 10 times difference in multi-core tests. It's not unreasonable to imagine someone plugging SRAM to an FPGA or ASIC and get 12ns access time where a PC needs 60. The cost of 96 MB of SRAM would certainly be prohibitively high though. I think that for what you guys are looking for, the algo mixes a lot of expensive features and makes it prohibitive to implement on hardware. I do have ideas how to help your algos be memory bound but they will be extremely slow and from what I'm reading the speed seems to be an issue for your use case.

**Comment by jdelorme3** [comment]
> Also looking at the numbers, it's said that an Nvidia 1080Ti does only 650 sol/s (=hashes/s) so it's even way lower than what we're doing on rfv2.

So this is what I was afraind of, equihash mostly target high performance hardware. I doubt I can run it on my Raspberry Pi!

**Comment by LinuXperia** [comment]
Ohh here it is already 2 AM need to go to sleep.

Look what it need to be done from my side of view is 
first calculate how many leading zeros are needed so we know how many bytes from the nonce endhash we need to map in the rambox on each rfv2_scanhash function call.

Here is a hint how to get the leading zeros from the nbits value that is in the pdata block header:
https://bitcoin.org/en/developer-reference#target-nbits

I will post a code example in next 24 hours.

All what is needed to do is extract the amount of the leading zeros from the nbits filed value
and after each RainForest Nonce endhash
 memcompare this amount of bytes in the rambox using the command
memcmp();

to avoid to much memory scan cost
als a random DAG like ethereum has it can be precomputed for x amount of Blocks and copied to the L1 or L2 cache and used it instead of the ram Box.

here you have more information about the DAG Ethereum uses.
we could also create a 10 Killobytes DAG load it after each endHash into L1 or L2 CPU Cash and scan the nonce Byte part there instead in the rambox.

https://www.sanfoundry.com/cpp-program-construct-random-directed-acyclic-graph/


So what we do after nBytes from the Nonce Endash was found in the specific memory is we submit it.

Okey have to go to sleep.
very late here.
Will write you back in 24 hours but i guess you will code it allready til then.

As said extract the amount of the leading zeros from the nbits then you know how many bytes from each Nonce endhash you need to map.

Then after each Nonce Endhash compare n Amount of leading bytes from the Nonce endhash and check if it exist in a Memory area.

If such a byte collusion was found then submit nonce to Network.

that is it.

In most case with this aproach we will find right Nonce afte Milli seconds of calculation so that way a Single Miner can Easy mine Millions of Blocks alone every 1 Minute as required.

the difficulty algorithm then will adjust the nbits vaule automaticly and by this require us to use more Bytes to check for collusion and by this slow us down to 1 Minute on average for 1 Block to Mine.

**Comment by bschn2** [comment]
But the nBits field is what defines the difficulty and the target. I'm not seeing how you want to mine faster when this difficulty is not changed. By the way @itwysgsl what's your lowest difficulty or nBits ? Bitcoin's minimal one is 0x1d00ffff which means a target of 0x00000000ffff00...00. It is chosen for very fast ASIC-compatible hashes (sha256) and long block life as it requires 4 billion hashes to scan entirely. Maybe you're still using the same ? The article pointed to above indicates that in regtest mode Bitcoin uses 0x207fffff which is 0x7fffff00...00, so if you forked off Bitcoin you also have access to this possible difficulty. At 1.1kH/s and a block per minute I'd suggest starting with nBits=0x1f00ffff for a difficulty of 1, which will give a target of 0x0000ffff00...00, hence exactly one block per minute.

**Comment by itwysgsl** [comment]
Hello @bschn2 I set min difficulty to `0x000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffff`
on test network for sake of easier testing.

**Comment by bschn2** [comment]
So that's the issue, it takes 256 minutes for a block at 1kH/s, thus it is only compatible with very fast hashes. Could you please change the difficulty to 0x0000ffff... and retest ? (no more than 4 zero digits on the left). It must work there. I'm asking this because I'm really concerned with weakening the hash. If we hash too fast, we take the risk of seeing strong optimizations come back, which is something we don't want to face anymore!

**Comment by itwysgsl** [comment]
@bschn2 I'm not sure if it's good idea to change min difficulty. It's fine for testnet but not for main network I think.

**Comment by MikeMurdo** [comment]
I've been following this thread for a while and am seeing something wrong here. Rfv2 has sensibly the same hash rate as scrypt, and requires a comparable difficulty. Plus if MBC emits a block each and every  minute this must be taken into account as well. With a target 0x00000000ffff... you need 4 billion hashes to find the block. At 1kH/s/core this means 4.3 million seconds*core or 49 days*cores. If you want to find them in less than one minute, you need 71000 cores. This is possibly fine when the coin is popular but it can significantly hinder its growth in its early days.

Even with your test's min difficulty of 0x000000ffff... you need 279 cores to find a block in one minute, which is verified by your experiments showing you didn't find a single block in 10 minutes with your i7.

Note that I'm not saying these numbers are not reasonable, and I don't buy the stupid paradigm of the solo miner who must be able to find a block even if he's alone : a coin having a single miner is deemed to die anyway. But I find it important to adapt the configuration to the algorithm's performance and not the opposite, or history will repeat itself.

We've seen the effect of the optimization of the last round in v1. I'm with @endlessloop2 here, please don't reduce the number of rounds! If it were just me I'd even add more to discourage implementations from even trying to unroll part of a loop (but I can admit that when there's less than 1% to shave it's unlikely they'll go through this pain).

Based on the numbers shown here, I think that 0x000000ff... for mainnet and 0x0000ff... for testnet are way more suitable targets to maintain a safe algorithm that people can still mine.

**Comment by itwysgsl** [comment]
@bschn2 I've made little research (checked code of other coins) and realized that many of them changed min diff during algo switching. I beleive it's one of possible soultions of current speed issue. Going to give it a shot right now.

**Comment by itwysgsl** [comment]
Also, can you guys check this https://github.com/bschn2/rainforest/issues/16 issue?

**Comment by MikeMurdo** [comment]
Great news for the difficulty @itwysgsl ! Keep up the good work, many of us would like to see you rule the crypto world!

**Comment by wtarreau** [comment]
I've replied to #16 (it's about CRC32 implementation).

**Comment by LinuXperia** [comment]
In Case a much better solution is still required here is how i would improve the code without
to change the Rainforest Algorithm

First as said extract how many leading zeros aka bytes are needed from the nBits Feild Value
```
@@ -44,6 +46,24 @@ int scanhash_rfv2(int thr_id, struct work *work, uint32_t max_nonce, uint64_t *h
        for (int k=0; k < 19; k++)
                be32enc(&endiandata[k], pdata[k]);
 
+       // EXTRACT HOW MANY BYTES FOR COLLUSION FROM THE nBITS FIELD VALUE
+       // Subtract from 0x20(32) minus the actual Diff aka 0x1f = to get how many bytes to check for collusion
+       uint32_t nBytesCollusion = 0x20 - (endiandata[18] >> 24);

```
Then what you do is instead to check if we are below a specific target we just look for byte collision of the Hash in a Memory Space be it in the RamBox or a DAG that is created same on all devices

```
@@ -55,13 +75,19 @@ int scanhash_rfv2(int thr_id, struct work *work, uint32_t max_nonce, uint64_t *h
                be32enc(&endiandata[19], nonce);
                rfv2_hash(hash, endiandata, 80, rambox, NULL);
 
-               if (hash[7] <= Htarg && fulltest(hash, ptarget)) {
+               // SCAN TROUGH 4096 MEMORY TO CHECK FOR A BYTE COLLUSION WITH nBYtes FROM THE HASH
+               for(uint32_t nBytePos=0; nBytePos<(4096-nBytesCollusion); nBytePos++) {
+                       //CHECK BYTE FOR BYTE IF A nBYTE COLLUSION EXIST IN THE MEM SPACE
+                       // IF YES SUBMIT SHARE / BLOCK TO NETWORK AS SOLUTION WAS FOUND
+                       if( !memcmp(ByteCollusionSpace+nBytePos, hash, nBytesCollusion) ) {
				work_set_target_ratio(work, hash);
				pdata[19] = nonce;
				*hashes_done = pdata[19] - first_nonce;
				ret = 1;
				goto out;
+			}
+		}
```

**Comment by LinuXperia** [comment]
> @bschn2 I've made little research (checked code of other coins) and realized that many of them changed min diff during algo switching. I beleive it's one of possible soultions of current speed issue. Going to give it a shot right now.

@itwysgsl The problem is what happens after lets say some mined blocks a big miner jumps in and increases the difficulty and after this goes quickly away ?

another possibility would be that a miner finds a vulnerbility and by this gains a huge advance, mine some blocks, increases the difficulty and goes away.

Then Coin will be blocked as no hashpower exist for such a difficulty.

You are stuck with a high difficulty and the coin will be not able to process any other blocks!

We are not going away from the problem that bruteforcing hashes is not what RainForest stands for.

Rainforest need another algorithm for finding a Solution without this stupid brutefoce hash target meeting.

**Comment by itwysgsl** [comment]
@LinuXperia this is well known issue, and very smart guy @zawy12 (author of LWMA algo) came up with nice solution of this issue - [TSA](https://github.com/zawy12/difficulty-algorithms/issues/36) difficulty ajustment algo. Difficulty is slowly going down during mining of block itself.

**Comment by itwysgsl** [comment]
P.s. @LinuXperia your idea with scanning rambox is also interesting, if you can make some proof of concept code, we can give it a shot :D

**Comment by itwysgsl** [comment]
@wtarreau can you also check this issue https://github.com/bschn2/rainforest/issues/17 ?

**Comment by itwysgsl** [comment]
I finally finished implementation of RFv2 in cpuminer https://github.com/MicroBitcoinOrg/Cpuminer/commit/8897a982b462f1fa66bb25559431e14913da1eb5 :D

**Comment by wtarreau** [comment]
I see that you enabled it side-by-side with rfv1. In the patch @MikeMurdo did, he removed support for v1 and replaced it with v2. I don't think the difference is important though :-)

**Comment by MikeMurdo** [comment]
So are your performance issues solved ?

**Comment by wtarreau** [comment]
At least on my i7-6700k, I'm getting this, which is pretty close to rf_test:
`[2019-04-08 00:13:27] Total: 5.64 kH/s`
Note that I lowered the benchmark target a little bit because it was scrolling too fast.

**Comment by bschn2** [comment]
I was about to close this one, but figured that if any participant wants to share new measures on various devices, it might still be the best place, thus I'm leaving it open. Please don't forget that the updated code is in the master branch now.

**Comment by wtarreau** [comment]
Good idea. rfv2_test gives:
  - 4200/s on my i7-6700k at 4.4 GHz (4 cores, 8 threads)
  - 1400/s on my J4105 at 2.4 GHz (4 cores, 4 threads)
  - 5600/s on my NanoPI-Neo4 at 2x2.1+4x1.7 GHz (2xA72+4xA53 cores, 6 threads)
  - 6600/s on my NanoPI-Fire3 at 1.6 GHz (8 A53 cores, 8 threads)
  - 780/s on an Atom Z8350 at 1.68 GHz (4 cores, 4 threads)
  - 330/s on a rk3288 (armv7, 4 cores at 1.9 GHz).

**Comment by jdelorme3** [comment]
8000/s on my friend's ryzen 2700 at 16x4 GHz !!!

**Comment by itwysgsl** [comment]
Hello everyone, I'm setting up new testnet with fixed RFv2. Going to post some speed results soon.

**Comment by wtarreau** [comment]
On Sun, Apr 28, 2019 at 10:34:50AM -0700, iamstenman wrote:
> Hello everyone, I'm setting new testnet with fixed RFv2. Going to post some
> speed results soon.

Does this mean there is an ip:port to which to connect to get stuff to hash ?
I'm asking because when I tried sgminer on my intel GPU some time ago there
was no bench mode and I had to connect to something more or less random that
would provide work, but I don't really know if that could have been part of
the failures, so for sure a public server could help. It's just a question
our of curiosity, I don't need any such thing right now anyway.

Thanks,
Willy

**Comment by itwysgsl** [comment]
@wtarreau yes, exactly, It's public test network main purpose of which - test and tune software and etc.

P.s. Updated RFv2 produce around 2.5 Kh/s on my Macbook's i7.

**Comment by wtarreau** [comment]
On Sun, Apr 28, 2019 at 11:24:33AM -0700, iamstenman wrote:
> @wtarreau yes, exactly, It's public test network main purpose of which - test
> and tune software and etc.

OK, thanks for your explanation.

> P.s. Updated RFv2 produce around 2.5 Kh/s on my I7.

Ah, great!

Willy

**Comment by jdelorme3** [comment]
And 2200 on my Raspberry Pi 3B+! Almost as fast as the Macbook! Thank you guys for your amazing work!!

**Comment by bschn2** [comment]
Cc: @itwysgsl @djm34 for their implementations

Gentlemen,

I found another remaining issue which really had to be addressed.  @itwysgsl I'm aware it's late for your deployment but better fix late than too late.

I was looking at the reasons for significant speed variations and figured that we didn't modulate the rambox access probability based on the number of loops, so some hashes were performing few changes to the rambox and others were performing a lot. My concern was not as much the variations as the likelihood that some miners could to try to distribute hashes by arranging them based on the expected rambox changes and try to combine them to minimize the amount of memory needed.

I have addressed this so that the probability is adjusted based on the number of loops, and now the amount of changes per hash varies much less (about 750 to 1350) and the performance is much more stable. I have updated the C and CL code. I also updated the test vectors.

Many thanks for your understanding,
Bill.

**Comment by jdelorme3** [comment]
OK just tried, was worried about performance changes, and it remains the same, even more stabler, 2160 to 2180 on my Raspberry Pi 3B+. 2200 I reported was average, it was oscilating from 2100 to 2300 (10%) but now its better (1%). Excellent work man!

**Comment by bschn2** [comment]
Many thanks for checking, Sir!

**Comment by itwysgsl** [comment]
@bschn2 oh, how critical this issue is?

**Comment by bschn2** [comment]
Hello Sir! "critical" is not the term, "concerning" is most suitable. But it typically is an overlooked design issue which can result in weakening the hashing difficulty for some hashes and just like the previous CRC padding issue, who knows to what extent it will be exploited once software start to squeeze the last nanoseconds out of a hash? I definitely don't want to see people start to merge their ramboxes between many threads by precomputing parts of the hashes and distributing them accordingly. For your usage even if the schedule is tight I think it's less troublesome to address it now than what it can result in if you have to hard fork again in 3 months.

**Comment by itwysgsl** [comment]
I see, thanks for info @bschn2 !

**Comment by djm34** [comment]
hmm with current change I am getting 300Hash/s on gpu...  (sure there are probably ways to optimize but it isn't a lot)

**Comment by bschn2** [comment]
Hello Sir! This is not expected at all, it reminds me when the RFV2_RAMBOX_HIST value was too low in the past. Are you sure it's at least 1536 as in the reference implementation ? The average number of rounds didn't change here, we're only making sure that the situations with less rounds maintain a proportional rambox access frequency. Most iterations have a large number of loops and experience no change since the limit remained the same.

**Comment by itwysgsl** [comment]
@bschn2 nodes is updated and new release is published

**Comment by bschn2** [comment]
Well done, Sir! I feel better with this addressed.

**Comment by djm34** [comment]
It is probably due to my implementation. 
I don't overwrite the rambox, but change locally the "prev" array which I have modified it's use. 
The main problem is that it has to scan through all the change, to see if something has been modified. (however I couldn't managed to run in parallel too many rambox

**Comment by bschn2** [comment]
I guess you're speaking about the for() loop here : https://github.com/MicroBitcoinOrg/CudaMiner/blob/4bc1f0cd17aff6979a337d52adce5b249a7b17d6/rainforest/rainforest_function.h#L569

Indeed, it will ruin performance. The purpose of the rambox was exactly to be fast (and so fast that it's sensitive to memory latency) so any loop there will have a disastrous effect. I'm interested in seeing how this evolves over time: some implementations might for example decide to ruin per-thread performance doing hash-table lookups and support many threads and others might prefer to limit the number of threads and maximize their performance. Right now I'd suggest to stick to the latter approach which is the same as the CPU code's. It allows to run about 10 threads per GB of RAM, which means roughly 40 threads on a 4 GB graphics card. This is already way more than what a standard CPU will do, and should likely result in comparable to slightly better performance.

**Comment by djm34** [comment]
what are the performance on cpu btw ?

tbh, I am a bit reluctant to modify the rambox, because a second access to a same index doesn't occur so often

**Comment by bschn2** [comment]
On my SkyLake 4.0 GHz with 8 threads:
```
./rfv2_test -b -t 8
3953 hashes, 1.094 sec, 8 threads, 3611.771 H/s, 451.471 H/s/thread
3828 hashes, 1.000 sec, 8 threads, 3827.843 H/s, 478.480 H/s/thread
3808 hashes, 1.000 sec, 8 threads, 3807.912 H/s, 475.989 H/s/thread
3863 hashes, 1.000 sec, 8 threads, 3862.911 H/s, 482.864 H/s/thread
3831 hashes, 1.000 sec, 8 threads, 3830.923 H/s, 478.865 H/s/thread
```
@jdelorme3 mentioned around 2200 H/s on his Raspberry Pi.

**Comment by wtarreau** [comment]
Hi!

Some numbers from my i6700K at 4.4:
```
$ ./rfv2_test -b -t 8
4301 hashes, 1.128 sec, 8 threads, 3814.046 H/s, 476.756 H/s/thread
4217 hashes, 1.000 sec, 8 threads, 4216.827 H/s, 527.103 H/s/thread
4164 hashes, 1.000 sec, 8 threads, 4163.900 H/s, 520.488 H/s/thread
4236 hashes, 1.000 sec, 8 threads, 4235.898 H/s, 529.487 H/s/thread
4135 hashes, 1.000 sec, 8 threads, 4134.909 H/s, 516.864 H/s/thread
4162 hashes, 1.000 sec, 8 threads, 4161.892 H/s, 520.236 H/s/thread
4143 hashes, 1.000 sec, 8 threads, 4142.905 H/s, 517.863 H/s/thread
```

NanoPI M4 base (2xA72 at 2.0 + 4xA53 at 1.5, RAM defaults):
```
$ ./rfv2_test -b -t 6
4366 hashes, 1.578 sec, 6 threads, 2766.324 H/s, 461.054 H/s/thread
3874 hashes, 1.000 sec, 6 threads, 3872.885 H/s, 645.481 H/s/thread
3686 hashes, 1.000 sec, 6 threads, 3684.979 H/s, 614.163 H/s/thread
3675 hashes, 1.000 sec, 6 threads, 3673.755 H/s, 612.292 H/s/thread
3640 hashes, 1.000 sec, 6 threads, 3638.239 H/s, 606.373 H/s/thread
3727 hashes, 1.001 sec, 6 threads, 3725.085 H/s, 620.848 H/s/thread
```

NanoPI M4 tuned (2xA72 at 2.1 + 4xA53 at 1.7, RAM at 928 MHz):
```
$ ./rfv2_test -b -t 6
5452 hashes, 1.371 sec, 6 threads, 3977.141 H/s, 662.857 H/s/thread
5360 hashes, 1.000 sec, 6 threads, 5359.234 H/s, 893.206 H/s/thread
5361 hashes, 1.000 sec, 6 threads, 5360.480 H/s, 893.413 H/s/thread
5367 hashes, 1.000 sec, 6 threads, 5366.501 H/s, 894.417 H/s/thread
5388 hashes, 1.000 sec, 6 threads, 5387.467 H/s, 897.911 H/s/thread
5422 hashes, 1.000 sec, 6 threads, 5421.458 H/s, 903.576 H/s/thread
```

Nanopi Fire3 (8xA53 at 1.6 GHz):
```
$ ./rfv2_test -b -t 8
6061 hashes, 2.582 sec, 8 threads, 2347.529 H/s, 293.441 H/s/thread
6212 hashes, 1.004 sec, 8 threads, 6187.534 H/s, 773.442 H/s/thread
6248 hashes, 1.004 sec, 8 threads, 6223.356 H/s, 777.919 H/s/thread
6131 hashes, 1.004 sec, 8 threads, 6106.902 H/s, 763.363 H/s/thread
6223 hashes, 1.004 sec, 8 threads, 6198.497 H/s, 774.812 H/s/thread
6415 hashes, 1.004 sec, 8 threads, 6389.735 H/s, 798.717 H/s/thread
6174 hashes, 1.004 sec, 8 threads, 6149.696 H/s, 768.712 H/s/thread
```

MacchiatoBin (4xA72 at 2.0 GHz):
```
$ ./rfv2_test -b -t 4
4068 hashes, 1.115 sec, 4 threads, 3649.455 H/s, 912.364 H/s/thread
4473 hashes, 1.000 sec, 4 threads, 4472.544 H/s, 1118.136 H/s/thread
4435 hashes, 1.000 sec, 4 threads, 4434.698 H/s, 1108.675 H/s/thread
4525 hashes, 1.000 sec, 4 threads, 4524.267 H/s, 1131.067 H/s/thread
4437 hashes, 1.000 sec, 4 threads, 4436.299 H/s, 1109.075 H/s/thread
4524 hashes, 1.000 sec, 4 threads, 4523.290 H/s, 1130.822 H/s/thread
4494 hashes, 1.000 sec, 4 threads, 4493.286 H/s, 1123.321 H/s/thread
```

UP Board (Atom X5-8350 at 4x1.68 GHz):
```
$ ./rfv2_test -b -t 4
759 hashes, 1.320 sec, 4 threads, 574.814 H/s, 143.704 H/s/thread
777 hashes, 1.000 sec, 4 threads, 776.880 H/s, 194.220 H/s/thread
758 hashes, 1.000 sec, 4 threads, 757.931 H/s, 189.483 H/s/thread
774 hashes, 1.000 sec, 4 threads, 773.927 H/s, 193.482 H/s/thread
760 hashes, 1.000 sec, 4 threads, 759.930 H/s, 189.983 H/s/thread
```

Hoping this helps!

**Comment by djm34** [comment]
ok around 18kh/s with some corners cut... (gives from time to time a non validated result)

**Comment by wtarreau** [comment]
On Fri, May 03, 2019 at 06:19:01AM -0700, djm34 wrote:
> ok around 18kh/s with some corners cut... (gives from time to time a non validated result)

That's impressive! What hardware is this ?

Willy

**Comment by djm34** [comment]
1080ti (this is pretty much the limit of the current implementation) with 2^13 thread

well probably not that impressive considering the hardware... a cpu with 32/64 thread will beat that easily

**Comment by wtarreau** [comment]
On Fri, May 03, 2019 at 06:57:14AM -0700, djm34 wrote:
> 1080ti (this is pretty much the limit of the current implementation) with 2^13 thread

Wow! This must be a monster! How do you deal with the rambox with so
many threads?

Willy

**Comment by djm34** [comment]
the rambox is common to all threads. And locally I match the rambox index to the ctx->changes index and passes and read changes to the rambox through the ctx.prev[]
the local indexing registers only rambox_idx/16 (assuming that it is rare to have consecutive index access to the rambox... which is in most cases true) in to uint16_t which allows to run more thread 

I saw actually some implementation of hashtable which should do a better job (assuming it works) than my fast implementation

**Comment by bschn2** [comment]
Well, this sounds like a really great optimization, Sir! I'm really glad you manage to maintain a high enough performance level on this GPU without them being excessively high, it preserves the users' investment while offering great opportunities for newcomers to fairly compete with these setups using low power devices.

**Comment by jdelorme3** [comment]
Hehe, 18k on a 1080Ti, this is only 9 times more higher than my Raspberry Pi 3B+! With 9 boards only I have the same performance with only 30 watts! MBC will be the best coin to mine :-)

**Comment by bschn2** [comment]
@jdelorme3 the CUDA implementation is still fairly recent and needs to workaround some tradeoffs imposed into the rambox to limit the scalability. But history and experience tell us that such optimizations will come and this performance will increase. With this said, your point is currently valid and it was my very first goal when working on this since last year. Further, look at @wtarreau 's numbers on the Nanopi Fire3. This board costs the same as your board (around $35) and is about 3 times faster. It is an important point to consider when building a new setup as it will further save energy and reduce electronic waste. And mining on your smartphone could be even twice as fast depending on the number of cores and their frequency.

**Comment by wtarreau** [comment]
FYI in PR #34 I fixed the time measurement causing the low value for the first line in bench mode that can be observed above in my test reports. It only affects the startup time, so the performance reported on subsequent lines is correct.

**Comment by itwysgsl** [comment]
@bschn2 @wtarreau we successfully switched to rfv2 few minutes ago!

**Comment by wtarreau** [comment]
On Tue, May 07, 2019 at 02:24:00AM -0700, iamstenman wrote:
> @bschn2 @wtarreau we successfully switched to rfv2 few minutes ago!

Oh, congratulations!

Willy

**Comment by itwysgsl** [comment]
Little update on situation after algo switch: at this moment we again have around 98% of hashrate from unknown source, and I have no idea why :(

**Comment by wtarreau** [comment]
On Sun, May 12, 2019 at 03:25:01AM -0700, iamstenman wrote:
> Little update on situation after algo switch: at this moment we again have
> around 98% of hashrate from unknown source, and I have no idea why :(

Just for my understanding, what does this mean exactly ? Is it that
someone is doing most of the work ? If you know it's an unknown source,
I'm assuming your can see the source(s), are you able to know if this
comes from multiple sources or a single one ? Could this mean that for
example there are too few people mining and that the only ones present
are the one with available hardware to throw at this task and are
collecting all the results ?

Willy

**Comment by itwysgsl** [comment]
@wtarreau I consider all hashrate comming not from public pools as "unknown source". There is [website](https://miningpoolstats.stream/microbitcoin) which allow monitor that. I'm not saying it's bad or good thing (or fault) of RFv2 (I'm pretty sure algo is fine) but just letting you know how things going on.

**Comment by wtarreau** [comment]
On Sun, May 12, 2019 at 04:10:19AM -0700, iamstenman wrote:
> @wtarreau I consider all hashrate comming not from public pools as "unknown
> source". There is [website](https://miningpoolstats.stream/microbitcoin)
> which allow monitor that. I'm not saying it's bad or good thing (or fault) of
> RFv2 (I'm pretty sure algo is fine) but just letting you know how things
> going on. 

OK I see. How is the global rate computed if some pools are not public ?
Also, is there any benefit for not going through the public ones ? For
example let's imagine that a huge mining farm made of big hardware like
you said was common doesn't want to pay a fee to a pool, it probably makes
sense not to use them and save a few percent, especially if this small
difference makes the difference with the electricity and cooling bill ?
Sorry for looking a bit dumb here, just trying to understand motivations
behind technical choices.

I'm seeing that the reported global hashrate (47 MH/s) roughly corresponds
to what can be achieved with 10000 smartphones. Thus if you only had 10k
users willing to run an application this could rebalance better.

Willy

**Comment by itwysgsl** [comment]
@wtarreau yeah, probably it's just big farm (it's most obvious explanation at the moment).

**Comment by MikeMurdo** [comment]
Few hours ago I submitted a PR for cpuminer-multi to integrate the new patches. It should boost adoption. GPUs tend to get most of the hashrate when a coin opens because big farms are readily available and you only need one of them picking the existing implementation and deploying massively to hundreds of nodes, while CPU ones take time to be noticed and adopted by the public. If GPUs are too strong initially it can scare CPU users who think they'll get too small shares. But if only 10k smartphones are needed to rebalance this, it will take time but will eventually work. How about we try to get the AMD miner up and running by the way ?

**Comment by itwysgsl** [comment]
@MikeMurdo working on sgminer right now, but not much is done by now (I don't even have AMD gpu for tests 😄).

**Comment by jdelorme3** [comment]
@MikeMurdo nice, once merged I can write an other article showing how to mine MBC with RaspberryPi 3B+ and mainline cpuminer. So much power at this cost, everyone should mine, at least to pay electricity and a bit more!

**Comment by bschn2** [comment]
Hello!
@MikeMurdo is right about the risk of initial strong start from GPUs, eventhough that's always a stressful period. However, what is wrong with the sgminer implementation? Doesn't work properly? I concede I didn't spend as much time on it as the rest of the code but did my best to keep it up to date.

**Comment by itwysgsl** [comment]
@bschn2 main issue with AMD miner development - lack of experience and AMD gpu for tests 😅

**Comment by bschn2** [comment]
Big shiver down my spine now : looking at the sgminer patch, it's the old one for RFv1!!! I don't know where I've put all the tedious work I did for RFv2, I hope it's not lost as I experienced lots of difficulties allocating private 96 MB areas to each thread, I hope I won't have to do it again! What scares me is that on my son's PC I boot on a USB stick and develop in /tmp which is mounted into RAM. I must have placed this code somewhere. I hope it didn't get lost when renaming the branches.

**Comment by itwysgsl** [comment]
@bschn2 just in case, you can check whole history of `rfv2.cl` file changes here https://github.com/bschn2/rainforest/commits/master/rfv2.cl

**Comment by wtarreau** [comment]
On Sun, May 12, 2019 at 04:48:25AM -0700, bschn2 wrote:
> I hope it didn't get lost when renaming the branches.

There is no such reason, and in any case your old work is still present.
If you want to recheck all your branches, you can :
  - run "git branch"
  - run "git log -p <branch>" on each of them to see each individual patch
    (hint: type '/^commit' to easily skip to next commit)
  - run "git reflog" to view all switches between branches, rebases etc.
    If you've ever committed that code, even if you switched to a different
    branch, or deleted that branch, it will be there. In this case I'd
    suggest doing "git log -p HEAD..<commit>" to visit what's in these
    respective branches. If it's there you'll find it.

Once you find the commit, do two things :
  - "git format-patch -k -1 <commit>" to make a copy of it
  - "git cherry-pick <commit>" to try to apply it (and possibly edit
    conflicts, then "git add -u", then "git cherry-pick --continue")

Good luck!
Willy

**Comment by itwysgsl** [comment]
Seams like I figured out source of unknown hashrate. While ago miners with tag `ccminer/jareso-experimental` (ccminer - Nvidia miner) joined to one of public pools and they have almost 5 MH/s per 8 GPUs. Probably rest of miners use same software.

In average 1 Nvidia GPU gets around 551.8 kH/s.

**Comment by wtarreau** [comment]
On Sun, May 12, 2019 at 06:44:27AM -0700, iamstenman wrote:
> Seams like I figured out source of unknown hashrate. While ago miners with
> tag `ccminer/jareso-experimental` (ccminer - Nvidia miner) joined to one of
> public pools and they have almost 5 MH/s per 8 GPUs. Probably rest of miners
> use same software.

Seems huge per GPU. Do you think these are the same types of GPUs with
3000+ cores as djm34's ? He saw ~18 kH/s on his device, which already
looked big to me. Also I'm wondering how they deal with the rambox with
so many threads.

Willy

**Comment by itwysgsl** [comment]
@wtarreau so far I have no details about this implemetation, but I also pretty sure it's using regular nvidia GPUs.

**Comment by bschn2** [comment]
Gentlemen,
I have uploaded a branch called "opencl-recovery" which contains an incomplete patch I found on my USB stick. It contains the C init code which performs the allocations and calls the OpenCL kernel. It still features the RFv1 C code, which I replaced last in the update process, so I take it as the backup I made after I saw the first valid hashes and before I replaced the C implementation. Also it does not contain the CL file which is a good hint that it's recent enough to rely on the official one that only needs be copied. I'm sharing this patch despite incomplete in hope it can help anyone (especially you, @itwysgsl), as I don't feel like doing it again for now I'm afraid. Please all accept my apologizes for not having being careful enough :(

**Comment by MikeMurdo** [comment]
@itwysgsl "In average 1 Nvidia GPU gets around 551.8 kH/s."
Seems a fairly high ratio to CPUs, very likely what they present as a single GPU is in fact a single CPU node with 5-7 GPUs connected to it. We've seen huge rigs on this coin in the past and its certain that the short block life encourages putting massive hardware at it to win the race ; if you cant manage to get a single share in one minute, you've wasted energy for nothing. On the opposite while there are few users this energy is well invested when you can make 90% of the shares. So its possible that your environment-friendly coin is responsible for a few tens of KW at its beginning, which is kinda amusing :)

**Comment by itwysgsl** [comment]
@MikeMurdo even if so, now we have >50 hashrate comming from public pools :)
Better than nothing.

**Comment by bschn2** [comment]
Gentlemen,

I must say I have been quite concerned over these last few days regarding these reports of high hash rates coming from nowhere on MBC, especially this experimental miner showing important values while being closed source, had it anything to hide.

So I decided to investigate. Today I've retrieved 9486 MBC block headers using mbc.wiki's API and decided to hash them again and see if a pattern would stand out. At least I can say my research was fruitful! Below I graphed the number of loops required to find each hash, from 2 to 257:
![strange](https://user-images.githubusercontent.com/39671501/57730112-0c997180-7698-11e9-998a-6671d33f8df6.png)
In a uniformly distributed space loops are about 95% of the time above 200. Here as you can all see, this only happened at the beginning then it stopped, and most hashes now only require 2 loops and are roughly 50 times faster to compute this way, while there are only 1/256 of them present overall. And the graph is becoming even sparser over time! This was made possible by the late drop of the floating point operations which turned transcendental functions to their discrete equivalents, which are thus easy to iterate over. Fortunately, our last minute change consisting in adjusting the rambox write ratio to be inversely proportional to the loops was present to save us!

My biggest concern is that this started with 2 loops from the very first 85 blocks! This means that someone figured it was faster to skip all other ones while we were still adjusting the code. This makes me very angry because we designed openly as a community and people watch us work hard, take notes of issues and keep them silently to exploit them later instead of reporting them while it's not too late. Oh they are certainly reading us right now...

This stopped at block 915085 and started again at 915328 and is still valid. A few blocks are still found by honest miners however, which is important.

So I have a solution to this which will not require to change the server code. I have modified the benchmark code of rfv2_test.c to behave like a scan function and to skip the nonces we're not intersted in. At first I saw that 50% of the time was spent computing CRCs on messages to skip them 255 times out of 256. By only changing the last word (like a nonce) and recalculating the CRC on the end, I can double the scan speed and spend 99% in the hashing function computing 2-loop hashes.
This is how my Skylake at 4 GHz now achieves 188 kH/s up from about 3600 without this. Yes it will be faster for everyone, not just for a small bunch of cheaters this time.

The next step is to retrofit this into the cpuminer's scan function. I'm now uploading the updated rfv2_test.c code to illustrate what is needed.

Please accept my apologizes for overlooking this design issue. In the end we're more or less back to the first step towards v2 which consisted in doubling the number of rounds and adding extra padding to be secure. We've still increased the rambox usage as well. But the extra protections like floats and variable number of loops are gone. This is still the benefit of having multiple protections though!

**Comment by bschn2** [comment]
I have now also uploaded the updated rfv2_cpuminer.c file. On my Skylake it shows 46MH/s but you have to take into account that 255 our of 256 such hashes are skipped thus the real effective rate is 1/256 of what is displayed, thus 180 kH/s.

**Comment by jdelorme3** [comment]
Many thanks Mr Schneider for your openess. Dont be sad, all coins and algos are attaqued. Your algo is resistant and it maintiens fairness between all users and that is much more important than what other algos do like sha256 who is mined only by big corporations using tons of asics.

I have just quickly tested rfv2_test on my Raspberry Pi 3B+ and it now say 47000 H/s. So I must divise by 256 that's it ?

**Comment by bschn2** [comment]
@jdelorme3 thank your for the kind words. No rfv2_test's numbers are real! It's only cpuminer which reports the hash interval as a hashrate because it expects hashes to be linearly explored. Pools however will report a performance level approximately matching rfv2_test.

47000 for such a device is quite not bad at all, our performance remains roughly flat between platforms under various conditions, this is reassuring. Now I'm doing roughly 4 times better than you while in the past I was doing twice. I would have been worried if we'd gone to a 1:10 or 1:100 factor, but here it's perfectly fine. I'd be curious about @wtarreau's farm with his more modern ARM cpus.

**Comment by jdelorme3** [comment]
Wow, 47000 real for my RPi, this is narly what some GPUs were doing on skypool a few days earlier!

**Comment by wtarreau** [comment]
OK guys, just two quick tests before going to bed
```
NanoPI-M4$ ./rfv2_test -b -t 6
91423 hashes, 1.000 sec, 6 threads, 91415.961 H/s, 15235.993 H/s/thread
91510 hashes, 1.000 sec, 6 threads, 91494.446 H/s, 15249.074 H/s/thread
91100 hashes, 1.000 sec, 6 threads, 91087.521 H/s, 15181.254 H/s/thread
91229 hashes, 1.000 sec, 6 threads, 91219.422 H/s, 15203.237 H/s/thread
91381 hashes, 1.000 sec, 6 threads, 91371.041 H/s, 15228.507 H/s/thread
91489 hashes, 1.000 sec, 6 threads, 91479.669 H/s, 15246.612 H/s/thread
91385 hashes, 1.000 sec, 6 threads, 91374.583 H/s, 15229.097 H/s/thread
```
```
NanoPI-Fire3$ ./rfv2_test -b -t 8
123646 hashes, 1.004 sec, 8 threads, 123146.150 H/s, 15393.269 H/s/thread
127629 hashes, 1.004 sec, 8 threads, 127123.050 H/s, 15890.381 H/s/thread
127779 hashes, 1.004 sec, 8 threads, 127272.582 H/s, 15909.073 H/s/thread
127883 hashes, 1.004 sec, 8 threads, 127375.663 H/s, 15921.958 H/s/thread
127838 hashes, 1.004 sec, 8 threads, 127330.334 H/s, 15916.292 H/s/thread
127931 hashes, 1.004 sec, 8 threads, 127424.615 H/s, 15928.077 H/s/thread
```
```
mcbin$ ./rfv2_test -b -t 4
81888 hashes, 1.000 sec, 4 threads, 81882.596 H/s, 20470.649 H/s/thread
89242 hashes, 1.000 sec, 4 threads, 89222.371 H/s, 22305.593 H/s/thread
89153 hashes, 1.000 sec, 4 threads, 89137.134 H/s, 22284.283 H/s/thread
89221 hashes, 1.000 sec, 4 threads, 89204.140 H/s, 22301.035 H/s/thread
89214 hashes, 1.000 sec, 4 threads, 89197.588 H/s, 22299.397 H/s/thread
89171 hashes, 1.000 sec, 4 threads, 89155.309 H/s, 22288.827 H/s/thread
```
So these were three in the end :-)

ARM are now slower than PCs however, likely due to the dominance of the rambox over expensive instructions. The results are still good.

However there is something I don't understand. If you skip 255 nonces every 256, how can you be sure you're not skipping the solution and that someone will actually find it ?

**Comment by bschn2** [comment]
Thanks for the numbers. @wtarreau there isn't a single solution but many. The probability that it's present in these nonces is the same as being present in others. We're looking for hash collisions in fact.

**Comment by djm34** [comment]
yes indeed, keeping only the loops = 2 allows many simplification such as there is no need to update the rambox. Applying these simplification gives hashrate around 200-300 kh/s (without any real optimization).

**Comment by wtarreau** [comment]
On Tue, May 14, 2019 at 07:00:51PM -0700, djm34 wrote:
> yes indeed, keeping only the loops = 2 allows many simplification such as
> there is no need to update the rambox.

Why would you not update the rambox in this case ? It's used inside each
round, so if you don't update it you will definitely produce bad hashes
from time to time if the same location is visited multiple times.

> Applying these simplification gives hashrate around 200-300 kh/s
> (without any real optimization)

Pretty good!

Willy

**Comment by wtarreau** [comment]
Just for fun I tried to optimize the code a little bit. My nanopi is now as fast as my laptop  (~+30%) and uses much less memory. I measured that the amount of changes made to the rambox remains low (180 per hash on avg), resulting in 1 chance out of 70000 to reuse a modified location. So I bypassed the changes and used a shared rambox to please my small cache. It now shows ~40MH/s which really equals 160k effective. I initially tried to compare the changes by hand but figured it's much more convenient to connect to a pool and see a share accepted!

However there is something very interesting in the variable rambox boundaries, which is that it's pointless to try to save and intermediate state because the whole hash completely changes depending on the nonce thanks to this. Thus in practice there are always two full rounds.

I currently have the patches in a local development branch applied directly on top of cpuminer, I'll rebase my rfv2 repo and will re-integrate them there later today.

**Comment by bschn2** [comment]
Excellent, Sir! I'm impatient to integrate your work!

**Comment by wtarreau** [comment]
It was sent now (PR #35)

**Comment by bschn2** [comment]
Yes we were talking past each other :-) Now merged, thank you!

**Comment by itwysgsl** [comment]
Hello @bschn2, big thanks to your work on this project, also thanks to @wtarreau for his supports.
I just builded cpuminer and keep getting ```Warning: rfv2_scan_hdr() returned invalid solution``` error messages. Since I'm using macOS I experienced some compilation issues related to lack of `htobe32` and `le32toh` functions (which I successfully copied from cpuminer code, here is all [code](https://gist.github.com/itwysgsl/100318b3d573085a56d0e22b94d7e222)).

Any thoughts on this issue? I suspect my macOS build fixes may break something so I going to test build on my Ubuntu VM.

**Comment by itwysgsl** [comment]
Ok, my bad. Now works like a charm with this implemetation https://gist.github.com/yinyin/2027912.

**Comment by bschn2** [comment]
Great news Sir!
Ideally we should rebuild the cpuminer patch set and try to get it merged upstream so that you don't need to maintain this fork anymore.

**Comment by djm34** [comment]
> On Tue, May 14, 2019 at 07:00:51PM -0700, djm34 wrote: yes indeed, keeping only the loops = 2 allows many simplification such as there is no need to update the rambox.
> Why would you not update the rambox in this case ? It's used inside each round, so if you don't update it you will definitely produce bad hashes from time to time if the same location is visited multiple times.

the probability of multiple reading is very low, especially in the 2 loops case. You have roughly 50 rounds over a huge rambox...

**Comment by itwysgsl** [comment]
@bschn2 since we added some platform specific functions like `be32toh` which present onlu on Linux, I think we should include [this](https://gist.github.com/panzi/6856583) header file to repository to make it comatible with all systems. I tested it with cpuminer and it works like a charm \:)

**Comment by wtarreau** [comment]
@itwysgsl we can do even simpler. I normally never use these macros, it's just that I lazily did that inside cpuminer which already uses them and didn't realize that by moving the code back to rfv2 I'd introduce this dependency. I'll replace them by generic functions. They're not even on a fast path, there's no reason to bother anyone with such dependencies.

**Comment by wtarreau** [comment]
PR #36 fixes this now.

**Comment by MikeMurdo** [comment]
> Ideally we should rebuild the cpuminer patch set and try to get it merged upstream so that you don't need to maintain this fork anymore.
Later today I should have some time to work on this and update my pending PR.

**Comment by MikeMurdo** [comment]
Still late on my work, I'm afraid it's not for today anymore. Wanted to let you know.

**Comment by itwysgsl** [comment]
Btw, yesterday @jareso (author of optimized private miner for rfv2) showed up in our discord and left this message: 
![](https://i.imgur.com/09PZKyB.png) 

Right now his miner, in average, produce 18 MH/s per miner.
Just want to let you know :)

**Comment by itwysgsl** [comment]
Also I tested RFv2 with recent patches on my android phone, and it produce around 10 kh/s.
https://github.com/itwysgsl/RainforestAndroidMiner/commit/ac744b3af58f239ea121d76157517b8538c5a0d2

**Comment by bschn2** [comment]
@itwysgsl his response is totally valid regarding the fact that others are attacking algorithms as well. What I'm contesting is not this. It's the fact that we're all working on a way to better decentralize mining, and that this person has the required skills to spot weaknesses **before** the algorithm is released, and silently keeps them secret incurring lots of work for you and the pools upon each release while disclosing such issues early would result in a more robust initial design. Thus I stand by my words, his primary motive is to grab most of the shares and not to help with fairness.

With this said, I really doubt he achieves 18 MH/s per card, I suspect there are quite a number of cards per miner; the algorithm involves operations that are extremely fast on ARM, very fast on x86 and not natively implemented on many other platforms, thus costing more. So the base performance per core will be lower. Anyway, this helps fuel my ideas for a v3 :)

I'm seeing on miningpoolstats that our latest update has done lots of good, with most miners going back to the public pools. Also there are mostly GPUs on zergpool and mostly CPUs on skypool. So far, so good.

Regarding your phone, it's ARMv7, right ? If so it's pretty decent for such a platform which doesn't have AES, IDIV, CRC nor 64-bit native operations! Let's wait for some feedback from users of more recent phones using ARMv8 (with CPUs like Snapdragon or Kryo).

**Comment by MikeMurdo** [comment]
@bschn2 the CPU vs GPU specialization of pools is natural, expected and nothing new : if you're having a low power (say a CPU) you'd rather join a pool where users are mostly like you so that you have a chance to get a respectable share every time a block's found, even if it's not frequent. But if you have 40 GPUs and can expect to help a pool mine a block 3 times more frequently and get 50% of the shares, you'd rather join a pool already showing a high hash rate because you'll get a big share frequently. And if you know you're concentrating most of the power, you'd rather join the pool with the lowest fees or run solo.

And with few miners at the moment, MBC is extremely appealing to mine : even with moderate power you can expect to make a good share of 12.5k every 2 minutes or so. I watched yesterday and estimated that with only 1 MH/s it was possible to make roughly $25/day. This is why those with large enough power immediately jump onto such coins, they are extremely profitable at opening. Once more miners join, the revenue is more evenly spread and it's less possible to expect such high revenues. At the moment it seems to go mostly to Jareso though.

@bschn2: most android phones with armv8 seem to enable armv7a only: https://i.stack.imgur.com/7EF24.jpg https://i.stack.imgur.com/XCGxi.jpg
However as you can see, aes, idiv and crc are present there and I think you may be able to enable them even in armv7 mode. I remember seeing a raspi somewhere in armv7 mode using CRC32 instructions. Possibly that building with "-march=armv7-a+idiv+crypto+crc" would enable everything. Or maybe with "-m32 -mcpu=cortex-a53+crypto+crc".

**Comment by itwysgsl** [comment]
@bschn2 I believe my phone have `Snapdragon 835` inside. Also to make android miner work I commented out [this](https://github.com/itwysgsl/RainforestAndroidMiner/blob/master/bitzenymininglibrary/src/main/cpp/libcpuminer/rfv2_core.c#L265-L289) part, otherwise error related to `__builtin_clrsbll` pop up.

**Comment by wtarreau** [comment]
It could be indicative of an arch mismatch indeed. Have you tried to pass "-march=armv8-a+crypto+crc" for example when building, instead of "-march=native" ?

**Comment by itwysgsl** [comment]
@wtarreau nope, but I can try right now.

**Comment by itwysgsl** [comment]
Here is error message itself `error: definition of builtin function '__builtin_clrsbll'
static`. Probably [this](https://github.com/bschn2/rainforest/blob/master/rfv2_core.c#L265-L266) check don't have appropriate condition to handle android build 🤔

**Comment by wtarreau** [comment]
Can you please report the output of "gcc -v" ? I guess the test on the version is not correct indeed. I did this part myself to try to build on older compilers. In the worst case you can simply disable this block and see if your perf is better with the build options.

**Comment by itwysgsl** [comment]
@wtarreau I think Android Studio uses clang for compilation, but still here is output for `gcc -v`:
```Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 10.0.0 (clang-1000.10.44.2)
Target: x86_64-apple-darwin18.2.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

**Comment by wtarreau** [comment]
Ah, got it! It's clang indeed, so it advertises v4.2.1. We'd need to check for the clang version then, as your version appears to include this builtin while the one I tested last time didn't. In the mean time it's harmless, feel free to comment out the whole block to be able to retest the performance on armv8 mode with all extensions enabled.

**Comment by itwysgsl** [comment]
@wtarreau so, with `-march=armv8-a+crypto+crc` flag miner loses around 20-40% of performance.

**Comment by wtarreau** [comment]
Oh that's really funny, probably that certain architectural optimizations are lost. From what I've found, SD835 is an octa-A73 up to 2.45 GHz, it must be awesome! I'm having a hard time imagining that it can only be a compiler issue. Please try to use "-mcpu=cortex-a73+crypto+crc" instead, and add -m64 to be certain it builds in 64-bit mode. Also double-check that you didn't lose the optimization level (typically -O3/-Ofast) when forcing the other flags.

**Comment by MikeMurdo** [comment]
I've updated the cpuminer PR with the latest speedups and fixes.

**Comment by itwysgsl** [comment]
@wtarreau with `-mcpu=cortex-a73+crypto+crc` I'm getting around 14 kh/s :)
Can't compile with `-m64` flag yet since it causes some troubles with sha256 implemetation in cpuminer. Working on a fix now.

**Comment by itwysgsl** [comment]
So, I tried it, but with `-m64` flag miner lose speed and don't produce any shares.

**Comment by wtarreau** [comment]
Ah so at least it proves it's currently running in 32-bit which explains the lower performance. If it's cpuminer you're using, you should build it with -DNOASM. Last time I updated the code there I even added a script "build-linux-arm" or something like this because the default options didn't work for me.

**Comment by djm34** [comment]
> Btw, yesterday @jareso (author of optimized private miner for rfv2) showed up in our discord and left this message:
> ![](https://camo.githubusercontent.com/40ea6002da324aeeb903e9d9965c290db4cafc2a/68747470733a2f2f692e696d6775722e636f6d2f3039505a4b79422e706e67)
> 
> Right now his miner, in average, produce 18 MH/s per miner.
> Just want to let you know :)

LOL gpu dev on the brink of a burnout

**Comment by wtarreau** [comment]
Guys, I've reviewed some details of the algo and also the performance ratio mentioned above. I'm thinking that the main issue is that the whole rambox should be used to make sure there is no way to bypass it nor to precompute it. The problem is that memory controllers nowadays are extremely fast and prefetch tons of data on large systems (high-end CPUs, GPUs) but are dumb and unable to prefetch much on cheap hardware. However the cheap hardware possesses architectural optimizations (crc32, aes) that large ones partially possesses (aes for x86) and that GPUs don't have.

If we consider the instruction-equivalent latency, by carefully chaining all this it should be possible to make everyone run at a bounded speed. Let's take my Neo4's 180ns RAM access time as a reference. This device has fast CPU cores with all extensions enabled. If we focus on a 200ns target as the time for a single operation, it means this machine must consume 20 more ns doing things that other machines will be slower at. My skylake has a 65ns access time. It does have AES but lacks CRC32. Fine, let's figure the fastest CRC32 implementation, repeat it as needed to fill. A raspi 3B+ shows 160ns, and it does have CRC32 but lacks AES. So we can put maybe one or two AES calls so that AES+RAM+CRC still gives 200ns.

Note, the PC will be able to prefetch multiple accesses at once so multiple threads will benefit from this. But we don't need to read lots of data, the bandwidth is not interesting here. Let's say we use a 64 MB work area (26 bits). This can be turned to 24 bits by considering 32-bit words. An AES operation returns 128 bits. This can be sufficient to produce 5 memory addresses to fetch data from. You also want to always write so that there is no efficient way to keep a shared read-only copy of this.

From what I'm reading at a few places, GPUs like the 1080Ti mentioned above seem to feature many memory channels (11, or 12 minus one to be more precise):
  https://www.anandtech.com/show/11180/the-nvidia-geforce-gtx-1080-ti-review
  https://www.anandtech.com/show/11180/the-nvidia-geforce-gtx-1080-ti-review/3
  https://www.quora.com/Why-is-the-size-of-RAM-in-my-GPU-GTX-1080-Ti-not-in-the-power-of-2

The prefetch is a full 64 bytes cache-line like on many other CPUs. So a GPU could be 12 times faster than a PC or an SBC thanks to this. There's not much that can be done by adding extra instructions in the loop, as these could be amortized on the huge number of cores. However the memory size can limit the number of active cores. A GPU with 12 GB of RAM could only have 192 active threads with 64 MB of work area. But even with 192 cores at work, I expect they could perform well on AES or CRC. But in this case we're at 12 times the performance, not 100 times or so.

Am I overlooking anything ?

**Comment by itwysgsl** [comment]
@wtarreau nice research, what you think about creating separate issue with discussion for Rainforest improvements?

**Comment by wtarreau** [comment]
you're right, we're not that much in the speed area anymore :-)

**Comment by bschn2** [comment]
Hello!
Indeed we'd rather move such a discussion somewhere else. A quick note though, as this /is/ related to this issue, which is that manipulating large amounts of RAM does have an impact on the hash rate. The initialization time for your 64 MB rambox can take a while. Even if we ignore it, let's say you want to modify your 64 MB 32-bit at a time, with 200ns access time. This is 16M x .2 = 3.2 seconds only for total access time, this would result in 0.3 H/s in the best case, which is not practical. Even if you only write 1M times you only have 6% chance of hitting an already visited location.  By considering them as 128 bit locations instead, you'd have 4M locations. At 1 million changes it's 22% chance of collision and 5 H/s. We could also decide to write full cache lines (64 bytes) so that we do touch the whole rambox with one million writes but it's going to be expensive. Even 512k writes would result in 40% chances of collision only for 10 H/s. At 40% chance of collision, you can use shared memory and scale a lot. If you use 3000 threads and only 60% of them provide usable shares, it's still 1800 threads so the large memory doesn' t protect us here.

But let's open a new issue for this discussion.

**Comment by MikeMurdo** [comment]
There was a report of incorrect speed on cpuminer with my PR there : https://github.com/tpruvot/cpuminer-multi/pull/39. Is anybody aware of this ?

**Comment by wtarreau** [comment]
On Sat, May 25, 2019 at 02:37:38PM -0700, MikeMurdo wrote:
> There was a report of incorrect speed on cpuminer with my PR there :
> https://github.com/tpruvot/cpuminer-multi/pull/39. Is anybody aware of this ?

Yes! I mentioned it when I brought the speedup patches. The problem was
that the reported hashes count is used to compute how often to print the
stats, and if we reported the correct value, the stats output was reported
256 times faster. Now I figured how to fix this, I have a patch somewhere,
I'll send it eventually.

Willy

**Comment by wtarreau** [comment]
I finally uploaded my fixes in PR#39.

**Comment by MikeMurdo** [comment]
Perfect I'll take it and rebuild my patch thanks man

**Comment by mcreator-kp** [comment]
> Btw, yesterday @jareso (author of optimized private miner for rfv2) showed up in our discord and left this message:
> ![](https://camo.githubusercontent.com/40ea6002da324aeeb903e9d9965c290db4cafc2a/68747470733a2f2f692e696d6775722e636f6d2f3039505a4b79422e706e67)
> 
> Right now his miner, in average, produce 18 MH/s per miner.
> Just want to let you know :)

Do u have jareso-experimental miner?
